### ISSUE TITLE
'ç' character does not work anymore on "US, intl., with dead keys" keyboard layout

### ISSUE THREAD

**Issue by vitormmatos** [NixOS/nixpkgs#239415]
'ç' does not work anymore on "US, intl., with dead keys" keyboard layout

On US, intl., with dead keys keyboard layout, press the single quote char and c after.
'+c=ć
'+c=ç (old way)

It's a pain for people like me in Brazil that use an US keyboard.
This issue also happens on Pop!_OS, Vanilla OS and Garuda Linux on those same layouts.

There are workarounds on those systems like this one:

> sudo vim /usr/lib/x86_64-linux-gnu/gtk-3.0/3.0.0/immodules.cache

> sudo vim /usr/lib/x86_64-linux-gnu/gtk-2.0/2.10.0/immodules.cache

> On both, find the lines starting with "cedilla" "Cedilla" and add :en to the line. Something like this:

> "cedilla" "Cedilla" "gtk30" "/usr/share/locale" "az:ca:co:fr:gv:oc:pt:sq:tr:wa:en"

> Change the Compose file:

> sudo sed -i /usr/share/X11/locale/en_US.UTF-8/Compose -e 's/ć/ç/g' -e 's/Ć/Ç/g'

> Instruct the system to load the cedilla module:

> Add those lines to /etc/environment:

> GTK_IM_MODULE=cedilla
> QT_IM_MODULE=cedilla

> Reboot and you are done. 

Still, since Nix folder structure is quite different I have no idea how to make a workaround similar to this.

**Comment by roberth** [comment]
Hi @vitormmatos,
I've had to move your issue because this kind of thing is implemented here in the nixpkgs repo, in `nixos/modules`. Indeed the structure is quite different, as we generate our configuration from our own options that you can set in `configuration.nix`.
The NixOS manual has [a section about configuring keyboard layouts](https://nixos.org/manual/nixos/stable/index.html#custom-xkb-layouts), which might be helpful in your case.

**Comment by vitormmatos** [comment]
> Hi @vitormmatos, I've had to move your issue because this kind of thing is implemented here in the nixpkgs repo, in `nixos/modules`. Indeed the structure is quite different, as we generate our configuration from our own options that you can set in `configuration.nix`. The NixOS manual has [a section about configuring keyboard layouts](https://nixos.org/manual/nixos/stable/index.html#custom-xkb-layouts), which might be helpful in your case.

Thank you for moving the thread to this section.

Even though there may be a workaround for using 'ç', this is extremely counter intuitive, and since there are hundreds of keyboard layouts, why not having one that works with this character?

**Comment by roberth** [comment]
I gather that section of documentation was not helpful to you then. I think it should be possible to select a standard layout that addresses your needs, perhaps with [`services.xserver.layout`](https://search.nixos.org/options?show=services.xserver.layout&query=xserver.layout).
Arguably this should be easier to find, easier to link to and this should work out of the box, e.g. by using the "calamares" installer.

**Comment by FelipeMarcelino** [comment]
Any updates regarding c cedilla here? Brazillian :)

**Comment by Ivsucram** [comment]
Bump. Same issue here.

**Comment by roberth** [comment]
I can't really help with this, but maybe try [NixOS Portuguese](https://matrix.to/#/#pt:nixos.org) on matrix?

**Comment by lucasshiva** [comment]
The same thing happens in Arch Linux. I fix it by modifying the `.XCompose` file and remapping the key. 

In NixOS, since I'm using `home-manager`, I place the following somewhere in my `home.nix` file:
```nix
home.file.".XCompose".text = ''
  include "%L"

  <dead_acute> <C> : "Ç"
  <dead_acute> <c> : "ç"
'';
```

That was enough for me. However, the [Arch Wiki page](https://wiki.archlinux.org/title/Xorg/Keyboard_configuration#Key_combinations) says that we may need to configure the graphical environment variable `GTK_IM_MODULE=xim` and/or `XMODIFIERS="@im=none"`.

**Comment by vitormmatos** [comment]
> XCompose

Interesting, is there an alternative solution for Wayland users?

**Comment by OrcsBR** [comment]
Hey vitor, putting GTK_IM_MODULE = "cedilla" and QT_IM_MODULE = "cedilla" on my environment variables (environment.variables option on nixos) seemed to fix it for all apps except the ones chromium based (including Vivaldi where I am typing this now). It appears to be a chromium specific problem, as per this post on [Arch Forums](https://bbs.archlinux.org/viewtopic.php?pid=2211593#p2211593). It seems that chromium+wayland - unlike on X11 - doesn't open *any* gtk immodule. So there's that.

**Comment by tobiasBora** [comment]
On a different NixOs computer (installed maybe 1.5 years ago?), I was able to get cedilla with US intl (tried with KDE plasma 5, x11) with `AltGr+comma (,) + c`, as described in https://superuser.com/a/1076106/770292. Is it not enough or not working for you?

**Comment by glwbr** [comment]
> Hey Vitor, setting GTK_IM_MODULE = "cedilla" and QT_IM_MODULE = "cedilla" on my environment variables (environment.variables option on nixos) seemed to fix it on wayland for all apps except the ones chromium based (including Vivaldi where I am typing this now). It appears to be a chromium specific problem, as per this post on [Arch Forums](https://bbs.archlinux.org/viewtopic.php?pid=2211593#p2211593). It seems that chromium+wayland - unlike on X11 - doesn't open _any_ gtk immodule. So there's that.

Settings the envvars fixed for the mentioned apps but in terminal or terminal apps I still get ć

**Comment by glwbr** [comment]
> On a different NixOs computer (installed maybe 1.5 years ago?), I was able to get cedilla with US intl (tried with KDE plasma 5, x11) with `AltGr+comma (,) + c`, as described in https://superuser.com/a/1076106/770292. Is it not enough or not working for you? AltGr may just be the right Alt on your keyboard.

This behavior has changed; you used to be able to press Alt + Comma, but now it doesn’t work like that anymore.

**Comment by tobiasBora** [comment]
> This behavior has changed; you used to be able to press Alt + Comma, but now it doesn’t work like that anymore.

Really? When has this been changed? I tried today on my system that is [2.5 months old](https://github.com/NixOS/nixpkgs/commit/a73246e2eef4c6ed172979932bc80e1404ba2d56) and if `' + c` does produce `ć` as described by the OP, `AltGr + comma` gives me `ç` as expected. I tried it with Wayland in Konsole, Firefox & Chromium, no issue (this was configured via KDE's keyboard layout settings).

**Comment by glwbr** [comment]
> > This behavior has changed; you used to be able to press Alt + Comma, but now it doesn’t work like that anymore.
> 
> Really? When has this been changed? I tried today on my system that is [2.5 months old](https://github.com/NixOS/nixpkgs/commit/a73246e2eef4c6ed172979932bc80e1404ba2d56) and if `' + c` does produce `ć` as described by the OP, `AltGr + comma` gives me `ç` as expected. I tried it with Wayland in Konsole, Firefox & Chromium, no issue (this was configured via KDE's keyboard layout settings).

I'm not sure what caused the change. If you're using flakes, maybe part of your lockfile is stuck on an older version, or perhaps your desktop environment is handling this differently? When I was using i3, it worked with AltGr as you mentioned, but now it doesn’t. This isn’t something exclusive to NixOS, though on Ubuntu, it used to work with an AltGr variation as well. I'm not sure where to look to fix this; I thought setting the XKB options would be enough.

**Comment by thiagokokada** [comment]
I always used this in my configuration and seems to works fine:

https://github.com/thiagokokada/nix-configs/blob/master/modules/nixos/desktop/locale.nix#L12-L14

Keep in mind that I use `us-intl` instead and don't care too much about the dead keys (e.g.: Alt Gr).

**Comment by glwbr** [comment]
> us-intl

Yeah, maybe the locale settings are what fix it. Here’s my config: https://github.com/glwbr/dotfiles/blob/main/modules/system/locale.nix. I'll try adjusting mine based on yours and see if it works.

**Comment by glwbr** [comment]
> I always used this in my configuration and seems to works fine:
> 
> https://github.com/thiagokokada/nix-configs/blob/a00e8aa50e6e5786e1d533389c100aca41800417/modules/nixos/desktop/locale.nix#L12-L14
> 
> Keep in mind that I use `us-intl` instead and don't care too much about the dead keys (e.g.: Alt Gr).
> 
> Edit: this works fine for both X11 (i3) and Wayland (Hyprland/Sway).

Changing the `LC_CTYPE` worked for me `´ + c`
